File: DevExpress-Demo/CS/PropertyGridDemo.Wpf/ViewModels/TabbedViewModule/Model/BorderOptions.cs
using DevExpress.Mvvm.POCO;
using DevExpress.Xpf.Core;
using System;
using System.Collections.Generic;
using System.ComponentModel.DataAnnotations;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using System.Windows.Media;

namespace PropertyGridDemo {
    public enum BorderOptionsType {
        Empty,
        Solid
    }    
    public static class BorderOptionsExtensions {
        public static BorderOptions CreateBorderOptions(this BorderOptionsType type, SeriesOptions options) {
            switch (type) {
                case BorderOptionsType.Solid:
                    return ViewModelSource.Factory<SeriesOptions, SolidBorderOptions>(x => new SolidBorderOptions(x))(options);
            }
            return ViewModelSource.Factory<SeriesOptions, BorderOptions>(x => new BorderOptions(x))(options);
        }
    }
    [MetadataType(typeof(Metadata.BorderOptionsDataMetadata))]
    public class BorderOptionsData {
        public BorderOptionsData() {
            Color = Colors.Transparent;
            Opacity = 0d;
            Thickness = 0d;
            DashStyle = DashStyles.Solid;
        }
        public Color Color { get; set; }
        public double Opacity { get; set; }
        public double Thickness { get; set; }
        public virtual DashStyle DashStyle { get; set; }
    }
    [MetadataType(typeof(Metadata.BorderOptionsMetadata))]
    public class BorderOptions : BaseOptions {
        public BorderOptions() { }
        public BorderOptionsType BorderType { get; set; }
        public BorderOptions(SeriesOptions root) : base(root) { }
        public virtual BorderOptionsData Data { get; set; }
    }
    [MetadataType(typeof(Metadata.SolidBorderOptionsMetadata))]
    public class SolidBorderOptions : BorderOptions {
        public SolidBorderOptions() : this(null) { }
        public SolidBorderOptions(SeriesOptions root) : base(root) {
            Color = Colors.Black;
            Opacity = 1d;
            Thickness = 0d;
            DashStyle = DashStyles.Solid;

        }
        public virtual Color Color { get; set; }
        public virtual double Opacity { get; set; }
        public virtual double Thickness { get; set; }
        public virtual DashStyle DashStyle { get; set; }
        protected void OnColorChanged() {
            Opacity = (double)Color.A / 255;
            UpdateData();
        }
        protected void OnOpacityChanged() {
            Color = new Color() { A = (byte)(Opacity * 255), R = Color.R, G = Color.G, B = Color.B };
            UpdateData();
        }
        protected void OnThicknessChanged() {
            UpdateData();
        }
        protected void OnDashStyleChanged() {
            UpdateData();
        }

        protected void UpdateData() {
            Data = new BorderOptionsData() {
                Color = this.Color,
                Opacity = this.Opacity,
                Thickness = this.Thickness,
                DashStyle = this.DashStyle
            };
        }
    }
}

File: DevExpress-Demo/VB/PropertyGridDemo.Wpf/ViewModels/TabbedViewModule/Model/BorderOptions.vb
Imports DevExpress.Mvvm.POCO
Imports DevExpress.Xpf.Core
Imports System
Imports System.Collections.Generic
Imports System.ComponentModel.DataAnnotations
Imports System.Linq
Imports System.Text
Imports System.Threading.Tasks
Imports System.Windows.Media
Imports System.Runtime.CompilerServices

Namespace PropertyGridDemo

    Public Enum BorderOptionsType
        Empty
        Solid
    End Enum

    Public Module BorderOptionsExtensions

        <Extension()>
        Public Function CreateBorderOptions(ByVal type As PropertyGridDemo.BorderOptionsType, ByVal options As PropertyGridDemo.SeriesOptions) As BorderOptions
            Select Case type
                Case PropertyGridDemo.BorderOptionsType.Solid
                    Return DevExpress.Mvvm.POCO.ViewModelSource.Factory(Of PropertyGridDemo.SeriesOptions, PropertyGridDemo.SolidBorderOptions)(Function(x) New PropertyGridDemo.SolidBorderOptions(x))(options)
            End Select

            Return DevExpress.Mvvm.POCO.ViewModelSource.Factory(Of PropertyGridDemo.SeriesOptions, PropertyGridDemo.BorderOptions)(Function(x) New PropertyGridDemo.BorderOptions(x))(options)
        End Function
    End Module

    <System.ComponentModel.DataAnnotations.MetadataTypeAttribute(GetType(PropertyGridDemo.Metadata.BorderOptionsDataMetadata))>
    Public Class BorderOptionsData

        Public Sub New()
            Me.Color = System.Windows.Media.Colors.Transparent
            Me.Opacity = 0R
            Me.Thickness = 0R
            Me.DashStyle = System.Windows.Media.DashStyles.Solid
        End Sub

        Public Property Color As Color

        Public Property Opacity As Double

        Public Property Thickness As Double

        Public Overridable Property DashStyle As DashStyle
    End Class

    <System.ComponentModel.DataAnnotations.MetadataTypeAttribute(GetType(PropertyGridDemo.Metadata.BorderOptionsMetadata))>
    Public Class BorderOptions
        Inherits PropertyGridDemo.BaseOptions

        Public Sub New()
        End Sub

        Public Property BorderType As BorderOptionsType

        Public Sub New(ByVal root As PropertyGridDemo.SeriesOptions)
            MyBase.New(root)
        End Sub

        Public Overridable Property Data As BorderOptionsData
    End Class

    <System.ComponentModel.DataAnnotations.MetadataTypeAttribute(GetType(PropertyGridDemo.Metadata.SolidBorderOptionsMetadata))>
    Public Class SolidBorderOptions
        Inherits PropertyGridDemo.BorderOptions

        Public Sub New()
            Me.New(Nothing)
        End Sub

        Public Sub New(ByVal root As PropertyGridDemo.SeriesOptions)
            MyBase.New(root)
            Me.Color = System.Windows.Media.Colors.Black
            Me.Opacity = 1R
            Me.Thickness = 0R
            Me.DashStyle = System.Windows.Media.DashStyles.Solid
        End Sub

        Public Overridable Property Color As Color

        Public Overridable Property Opacity As Double

        Public Overridable Property Thickness As Double

        Public Overridable Property DashStyle As DashStyle

        Protected Sub OnColorChanged()
            Me.Opacity = CDbl(Me.Color.A) / 255
            Me.UpdateData()
        End Sub

        Protected Sub OnOpacityChanged()
            Me.Color = New System.Windows.Media.Color() With {.A = CByte((Me.Opacity * 255)), .R = Me.Color.R, .G = Me.Color.G, .B = Me.Color.B}
            Me.UpdateData()
        End Sub

        Protected Sub OnThicknessChanged()
            Me.UpdateData()
        End Sub

        Protected Sub OnDashStyleChanged()
            Me.UpdateData()
        End Sub

        Protected Sub UpdateData()
            Me.Data = New PropertyGridDemo.BorderOptionsData() With {.Color = Me.Color, .Opacity = Me.Opacity, .Thickness = Me.Thickness, .DashStyle = Me.DashStyle}
        End Sub
    End Class
End Namespace
